File: dmn-test-cases/standard/tck/1.3/cl3/0004-lending/translator/expected/java/dmn/type/Monthly.java
package type;

import java.util.*;

@javax.annotation.Generated(value = {"itemDefinitionInterface.ftl", "Monthly"})
@com.fasterxml.jackson.annotation.JsonPropertyOrder(alphabetic = true)
@com.fasterxml.jackson.databind.annotation.JsonDeserialize(as = type.MonthlyImpl.class)
public interface Monthly extends com.gs.dmn.runtime.DMNType {
    static Monthly toMonthly(Object other) {
        if (other == null) {
            return null;
        } else if (Monthly.class.isAssignableFrom(other.getClass())) {
            return (Monthly)other;
        } else if (other instanceof com.gs.dmn.runtime.Context) {
            MonthlyImpl result_ = new MonthlyImpl();
            Object income = ((com.gs.dmn.runtime.Context)other).get("Income");
            result_.setIncome((java.lang.Number)income);
            Object expenses = ((com.gs.dmn.runtime.Context)other).get("Expenses");
            result_.setExpenses((java.lang.Number)expenses);
            Object repayments = ((com.gs.dmn.runtime.Context)other).get("Repayments");
            result_.setRepayments((java.lang.Number)repayments);
            return result_;
        } else if (other instanceof com.gs.dmn.runtime.DMNType) {
            return toMonthly(((com.gs.dmn.runtime.DMNType)other).toContext());
        } else {
            throw new com.gs.dmn.runtime.DMNRuntimeException(String.format("Cannot convert '%s' to '%s'", other.getClass().getSimpleName(), Monthly.class.getSimpleName()));
        }
    }

    @com.fasterxml.jackson.annotation.JsonGetter("Income")
    java.lang.Number getIncome();

    @com.fasterxml.jackson.annotation.JsonGetter("Expenses")
    java.lang.Number getExpenses();

    @com.fasterxml.jackson.annotation.JsonGetter("Repayments")
    java.lang.Number getRepayments();

    default com.gs.dmn.runtime.Context toContext() {
        com.gs.dmn.runtime.Context context = new com.gs.dmn.runtime.Context();
        context.add("income", getIncome());
        context.add("expenses", getExpenses());
        context.add("repayments", getRepayments());
        return context;
    }

    default boolean equalTo(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Monthly other = (Monthly) o;
        if (this.getExpenses() != null ? !this.getExpenses().equals(other.getExpenses()) : other.getExpenses() != null) return false;
        if (this.getIncome() != null ? !this.getIncome().equals(other.getIncome()) : other.getIncome() != null) return false;
        if (this.getRepayments() != null ? !this.getRepayments().equals(other.getRepayments()) : other.getRepayments() != null) return false;

        return true;
    }

    default int hash() {
        int result = 0;
        result = 31 * result + (this.getExpenses() != null ? this.getExpenses().hashCode() : 0);
        result = 31 * result + (this.getIncome() != null ? this.getIncome().hashCode() : 0);
        result = 31 * result + (this.getRepayments() != null ? this.getRepayments().hashCode() : 0);
        return result;
    }

    default String asString() {
        StringBuilder result_ = new StringBuilder("{");
        result_.append("Expenses=" + getExpenses());
        result_.append(", Income=" + getIncome());
        result_.append(", Repayments=" + getRepayments());
        result_.append("}");
        return result_.toString();
    }
}

File: dmn-test-cases/proto/1.1/0004-lending/interpreter/expected/proto3/kotlin/dmn/type/Monthly.kt
package type

import java.util.*

@javax.annotation.Generated(value = ["itemDefinitionInterface.ftl", "Monthly"])
@com.fasterxml.jackson.annotation.JsonPropertyOrder(alphabetic = true)
@com.fasterxml.jackson.databind.annotation.JsonDeserialize(`as` = type.MonthlyImpl::class)
interface Monthly : com.gs.dmn.runtime.DMNType {
    @get:com.fasterxml.jackson.annotation.JsonGetter("Income")
    val income: java.math.BigDecimal?

    @get:com.fasterxml.jackson.annotation.JsonGetter("Expenses")
    val expenses: java.math.BigDecimal?

    @get:com.fasterxml.jackson.annotation.JsonGetter("Repayments")
    val repayments: java.math.BigDecimal?

    override fun toContext(): com.gs.dmn.runtime.Context {
        val context = com.gs.dmn.runtime.Context()
        context.put("income", this.income)
        context.put("expenses", this.expenses)
        context.put("repayments", this.repayments)
        return context
    }

    fun equalTo(o: Any?): Boolean {
        if (this === o) return true
        if (javaClass != o?.javaClass) return false

        val other = o as Monthly
        if (if (this.expenses != null) this.expenses != other.expenses else other.expenses != null) return false
        if (if (this.income != null) this.income != other.income else other.income != null) return false
        if (if (this.repayments != null) this.repayments != other.repayments else other.repayments != null) return false

        return true
    }

    fun hash(): Int {
        var result = 0
        result = 31 * result + (if (this.expenses != null) this.expenses.hashCode() else 0)
        result = 31 * result + (if (this.income != null) this.income.hashCode() else 0)
        result = 31 * result + (if (this.repayments != null) this.repayments.hashCode() else 0)
        return result
    }

    fun asString(): String {
        val result_ = StringBuilder("{")
        result_.append("Expenses=" + expenses)
        result_.append(", Income=" + income)
        result_.append(", Repayments=" + repayments)
        result_.append("}")
        return result_.toString()
    }

    companion object {
        @JvmStatic
        fun toMonthly(other: Any?): Monthly? {
            if (other == null) {
                return null
            } else if (other is Monthly?) {
                return other
            } else if (other is com.gs.dmn.runtime.Context) {
                var result_ = MonthlyImpl()
                result_.income = other.get("Income") as java.math.BigDecimal?
                result_.expenses = other.get("Expenses") as java.math.BigDecimal?
                result_.repayments = other.get("Repayments") as java.math.BigDecimal?
                return result_
            } else if (other is com.gs.dmn.runtime.DMNType) {
                return toMonthly(other.toContext())
            } else if (other is proto.Monthly) {
                var result_: MonthlyImpl = MonthlyImpl()
                result_.income = java.math.BigDecimal.valueOf((other as proto.Monthly).getIncome())
                result_.expenses = java.math.BigDecimal.valueOf((other as proto.Monthly).getExpenses())
                result_.repayments = java.math.BigDecimal.valueOf((other as proto.Monthly).getRepayments())
                return result_
            } else {
                throw com.gs.dmn.runtime.DMNRuntimeException(String.format("Cannot convert '%s' to '%s'", other.javaClass.getSimpleName(), Monthly::class.java.getSimpleName()))
            }
        }

        @JvmStatic
        fun toProto(other: Monthly?): proto.Monthly {
            var result_: proto.Monthly.Builder = proto.Monthly.newBuilder()
            if (other != null) {
                var incomeProto_: Double = (if ((other as Monthly).income == null) 0.0 else (other as Monthly).income!!.toDouble())
                result_.setIncome(incomeProto_)
                var expensesProto_: Double = (if ((other as Monthly).expenses == null) 0.0 else (other as Monthly).expenses!!.toDouble())
                result_.setExpenses(expensesProto_)
                var repaymentsProto_: Double = (if ((other as Monthly).repayments == null) 0.0 else (other as Monthly).repayments!!.toDouble())
                result_.setRepayments(repaymentsProto_)
            }
            return result_.build()
        }

        @JvmStatic
        fun toProto(other: List<Monthly?>?): List<proto.Monthly>? {
            if (other == null) {
                return null
            } else {
                return other.stream().map({o -> toProto(o)}).collect(java.util.stream.Collectors.toList())
            }
        }
    }
}
